SUPESC-676 changes in gift card feature installation (#1594)
spryker/spryker-docs: _includes/pbc/all/install-features/202304.0/install-the-gift-cards-feature.md

use Spryker\Zed\Payment\PaymentDependencyProvider as SprykerPaymentDependencyProvider;

class PaymentDependencyProvider extends SprykerPaymentDependencyProvider
{
    /**
     * @return \Spryker\Zed\Payment\Dependency\Plugin\Payment\PaymentMethodFilterPluginInterface[]
     */
    protected function getPaymentMethodFilterPlugins()
    {
        return [
            new PriceToPayPaymentMethodFilterPlugin(),
            new GiftCardPaymentMethodFilterPlugin(),
        ];
    }
}
```

**src/Pyz/Zed/SalesPayment/SalesPaymentDependencyProvider.php**

```php
<?php

namespace Pyz\Zed\SalesPayment;

use Spryker\Zed\GiftCard\Communication\Plugin\SalesPayment\GiftCardPaymentMapKeyBuilderStrategyPlugin;
use Spryker\Zed\SalesPayment\SalesPaymentDependencyProvider as SprykerSalesPaymentDependencyProvider;

class SalesPaymentDependencyProvider extends SprykerSalesPaymentDependencyProvider
{
    /**
     * @return array<\Spryker\Zed\SalesPaymentExtension\Dependency\Plugin\PaymentMapKeyBuilderStrategyPluginInterface>
     */
    protected function getPaymentMapKeyBuilderStrategyPlugins(): array
    {
        return [

>             new GiftCardPaymentMapKeyBuilderStrategyPlugin(),

Reviewer: please mentioned in list of plugins above

        ];
    }
}
```

{% info_block warningBox "Verification" %}

Make sure that:
* You can activate a gift card using its generated code.
* You can't activate a gift card the balance of which has been depleted.
* During the checkout process, payment method selection is skipped in case the gift card covers the grand total.

> * Having made a successful purchase with the help of a gift card, you receive a gift card balance notification e-mail.

Reviewer: add a verification point that when placing an order with more than one gift card

This document describes how to integrate the [Gift Cards feature](/docs/pbc/all/gift-cards/{{site.version}}/gift-cards.html) into a Spryker project.

## Install feature core

Follow the steps below to install the Gift Cards feature core.

### Prerequisites

To start feature integration, integrate the required features:

> | NAME   | VERSION | INTEGRATE GUIDE |

Reviewer 2: I added links to IGs. Make sure they are correct.

Author: @ilyakubanov I left all links to other IGs with old version 202212.0 confirm pls that it's right

Reviewer 2: @oleksandr-stoilovskyi you option is not correct. I replaced versions with {{site.version}}

|Product  | {{site.version}} |[Product feature integration](/docs/pbc/all/product-information-management/{{site.version}}/install-and-upgrade/install-features/install-the-product-feature.html)|
|Payments  | {{site.version}} |[Payments feature integration](/docs/scos/dev/feature-integration-guides/{{site.version}}/payments-feature-integration.html)|
| Shipment | {{site.version}} |[Integrate the Shipment feature](/docs/pbc/all/carrier-management/{{site.version}}/install-and-upgrade/integrate-the-shipment-feature.html)|
| Order Management | {{site.version}} |[Order Management feature integration](/docs/scos/dev/feature-integration-guides/{{site.version}}/order-management-feature-integration.html)|
| Mailing &amp; Notifications | {{site.version}} |[Mailing &amp; Notifications feature integration](/docs/scos/dev/feature-integration-guides/{{site.version}}/mailing-and-notifications-feature-integration.html)|
| Promotions &amp; Discounts | {{site.version}} |[Integrate the Promotions &amp; Discounts feature](/docs/pbc/all/discount-management/{{site.version}}/install-and-upgrade/integrate-the-promotions-and-discounts-feature.html)|

### 1) Install the required modules using Composer

> Install the required modules:

Reviewer 2: ```suggestion
```

```bash
composer require spryker-feature/gift-cards:"{{site.version}}" --update-with-dependencies
```

{% info_block warningBox "Verification" %}

Make sure that the following modules have been installed:

| MODULE | EXPECTED DIRECTORY |
| --- | --- |
| CartCode | vendor/spryker/cart-code |
| CartCodeExtension | vendor/spryker/cart-code-extension |
| GiftCard | vendor/spryker/gift-card |
| GiftCardBalance | vendor/spryker/gift-card-balance |
| Nopayment | vendor/spryker/nopayment |

{% endinfo_block %}

> ### 2) Set up configuration

Reviewer 2: As this section is blank and does not have any steps, just sub-steps, we can get rid of it
```suggestion
### 2) Set up the Gift Card purchase process
```

> #### Gift Card purchase process

Reviewer 2: ```suggestion
```

Extend your project with the following configuration.

**src/Pyz/Zed/GiftCard/GiftCardConfig.php**

```php
<?php

namespace Pyz\Zed\GiftCard;

use Spryker\Shared\Shipment\ShipmentConfig;
use Spryker\Zed\GiftCard\GiftCardConfig as SprykerGiftCardConfig;

class GiftCardConfig extends SprykerGiftCardConfig
{
    /**
     * Provides a list of shipment method names that should be available in case there are only gift card items in the quote.
     *
     * @return array
     */
    public function getGiftCardOnlyShipmentMethods(): array
    {
        return [
            ShipmentConfig::SHIPMENT_METHOD_NAME_NO_SHIPMENT,
        ];
    }
}
```

{% info_block warningBox "Verification" %}

When you finish the *Setup Behaviour* step, make sure that the `NoShipment` shipment method is selected automatically while ordering only a gift card.

{% endinfo_block %}

> #### Gift Card usage process

Reviewer 2: ```suggestion
3) ### Set up the Gift Card usage process
```

```php
<?php

namespace Pyz\Zed\GiftCard;

use Spryker\Shared\DummyPayment\DummyPaymentConfig;
use Spryker\Zed\GiftCard\GiftCardConfig as SprykerGiftCardConfig;

class GiftCardConfig extends SprykerGiftCardConfig
{
    /**
     * Provides a list of payment method names that are disabled to use when the quote contains gift card item(s) to purchase.
     *
     * @return array
     */
    public function getGiftCardPaymentMethodBlacklist(): array
    {
        return [
            DummyPaymentConfig::PAYMENT_METHOD_INVOICE,
        ];
    }
}
```

**src/Pyz/Zed/Sales/SalesConfig.php**

```php
<?php

namespace Pyz\Zed\Sales;

use Spryker\Zed\Sales\SalesConfig as SprykerSalesConfig;

class SalesConfig extends SprykerSalesConfig
{
    /**
     * @return array
     */
    public function getSalesDetailExternalBlocksUrls()
    {
        $projectExternalBlocks = [
            'giftCards' => '/gift-card/sales/list', // lists used gift cards for the order
        ];

        $externalBlocks = parent::getSalesDetailExternalBlocksUrls();

        return array_merge($externalBlocks, $projectExternalBlocks);
    }
}
```

{% info_block warningBox "Verification" %}

> When you finish the Setup Behaviour step, make sure the following:

Reviewer 2: Can you add the link to the step you are mentioning—the *Setup Behaviour step*

Author: @ilyakubanov  can you confirm my changes here
`[Setup Behaviour](#install-the-required-modules-using-composer)`
?

* The `NoPayment01` state machine is activated successfully.
* When using a gift card to cover an entire order, the configured order state machine is used—for example, `Nopayment01`.
* You can't use denylisted payment methods when using a gift card.
* In the Back office, on the order details page, you see the gift cards used in the order.

{% endinfo_block %}

> ### 3) Set up the database schema

Reviewer 2: ```suggestion
### 4) Set up the database schema
```

| src/Orm/Zed/GiftCard/Persistence/SpyGiftCardProductConfigurationLink.php | Spryker\\Zed\\GiftCard\\Persistence\\Propel\\AbstractSpyGiftCardProductConfigurationLink |
| src/Orm/Zed/GiftCard/Persistence/SpyGiftCardProductConfigurationLinkQuery.php | Spryker\\Zed\\GiftCard\\Persistence\\Propel\\AbstractSpyGiftCardProductConfigurationLinkQuery |
| src/Orm/Zed/GiftCard/Persistence/SpyPaymentGiftCard.php | Spryker\\Zed\\GiftCard\\Persistence\\Propel\\AbstractSpyPaymentGiftCard |
| src/Orm/Zed/GiftCard/Persistence/SpyPaymentGiftCardQuery.php | Spryker\\Zed\\GiftCard\\Persistence\\Propel\\AbstractSpyPaymentGiftCardQuery |
| src/Orm/Zed/GiftCardBalance/Persistence/SpyGiftCardBalanceLog.php | Orm\\Zed\\GiftCardBalance\\Persistence\\Base\\SpyGiftCardBalanceLog |
| src/Orm/Zed/GiftCardBalance/Persistence/SpyGiftCardBalanceLogQuery.php | Orm\\Zed\\GiftCardBalance\\Persistence\\Base\\SpyGiftCardBalanceLogQuery |
| src/Orm/Zed/Sales/Persistence/SpySalesOrderItemGiftCard.php | Spryker\\Zed\\Sales\\Persistence\\Propel\\AbstractSpySalesOrderItemGiftCard |
| Spryker\\Zed\\Sales\\Persistence\\Propel\\AbstractSpySalesOrderItemGiftCardQuery |

{% endinfo_block %}

> ### 4) Import data

Reviewer 2: This section is blank and does not provide any extra information, therefore we can remove it and replace it with the subsection. 
```suggestion
### 5) Import the Gift Card configuration data
```

> #### Gift Card configuration data

Reviewer 2: ```suggestion
```

1234,{prefix}-{randomPart}-{suffix}
```

| COLUMN | REQUIRED? | DATA TYPE | DATA EXAMPLE | DATA EXPLANATION |
| --- | --- | --- | --- | --- |
| abstract_sku |mandatory  | string | 1234 |  SKU reference of an abstract gift card product.|
| pattern |mandatory  | string | {prefix}-{randomPart}-{suffix} | A pattern that is used to generate codes for purchased gift card codes. |

**data/import/gift_card_concrete_configuration.csv**

```yaml
sku,value
1234_1,1000
1234_2,2000
1234_3,5000
```

| COLUMN | REQUIRED? | DATA TYPE | DATA EXAMPLE | DATA EXPLANATION |
| --- | --- | --- | --- | --- |
| sku | mandatory | string| 1234 | An SKU reference of an abstract gift card product. |
| value | mandatory | string	 | {prefix}-{randomPart}-{suffix} | A pattern that is used to generate codes for purchased gift card codes. |

```bash
console data:import:gift-card-abstract-configuration
console data:import:gift-card-concrete-configuration
```

{% info_block warningBox "Verification" %}

Make sure to have imported abstract and concrete gift card configuration into your `spy_gift_card_product_abstract_configuration` and `spy_gift_card_product_configuration` database tables.

{% endinfo_block %}

> #### Shipment method data

Reviewer 2: ```suggestion
### 6) Set up the shipment method data
```

{% info_block infoBox "Info" %}

In this step, you create a shipment method called `NoShipment`. The name of the shipment method has to match the value of `\Spryker\Shared\Shipment\ShipmentConfig::SHIPMENT_METHOD_NAME_NO_SHIPMENT` constant.

{% endinfo_block %}

1. Taking into account project customizations, extend shipment method data importer as shown below:

**data/import/shipment.csv**

```yaml
shipment_method_key,name,carrier,taxSetName
spryker_no_shipment,NoShipment,NoShipment,Tax Exempt
```

2. Taking into account project customizations, extend your shipment price data importer as shown below:

**data/import/shipment_price.csv**

```yaml
shipment_method_key,store,currency,value_net,value_gross
spryker_no_shipment,DE,EUR,0,0
```

3. Apply changes:

```bash
console data:import:shipment
console data:import:shipment-price
```

{% info_block warningBox "Verification" %}

Make sure that a shipment method with the `NoShipment` name exists in your `spy_shipment_method` and `spy_shipment_method_price` database tables.

{% endinfo_block %}

> #### Additional, optional data imports

Reviewer 2: ```suggestion
### 7) Import additional and optional data
```

{% info_block infoBox "Info" %}

To represent and display gift cards as products in your shop, you need to import some data into your database depending on your project configuration and needs. The following list contains the points that can be used to get an idea of what gift card-related data you might want to use:
* *Product Attribute Key* to create a gift card `value` super attribute that defines gift card variants.
* *Abstract Product* that represents gift cards in your catalog.
* *Abstract Product Store Relation* to manage store-specific gift cards.
* *Concrete Product* that represents gift cards with a specific price value.
* *Product Image* for abstract and concrete products to display gift cards.
* *Product Price* for concrete gift card products where the price value matches the "value" super attribute.
* *Product Stock* data for concrete gift card products.
* *Product Management Attribute* to define the previously created "value" product attribute for the PIM.
* *Category* that represents all gift cards.
* *Navigation item* to display the gift card category or gift card product details page directly.

{% endinfo_block %}

> ### 5) Prepare order state machines: Gift Card purchase process

Reviewer 2: Can we adjust the title as follows?
```suggestion
### 8) Prepare order state machines for the Gift Card purchase process
```

```html
<?xml version="1.0"?>
<statemachine
        xmlns="spryker:oms-01"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="spryker:oms-01 http://static.spryker.com/oms-01.xsd">

    <process name="DummyRefund">

        <states>
            <state name="ready for return" />
            <state name="returned"/>
            <state name="refunded"/>
        </states>

        <transitions>
            <transition>
                <source>ready for return</source>
                <target>returned</target>
                <event>execute-return</event>
            </transition>

            <transition>
                <source>returned</source>
                <target>refunded</target>
                <event>refund</event>
            </transition>
        </transitions>

        <events>
            <event name="execute-return" onEnter="true"/>
            <event name="refund" manual="true" command="DummyPayment/Refund"/>
        </events>
    </process>

</statemachine>
```

![Nopayment](https://spryker.s3.eu-central-1.amazonaws.com/docs/Migration+and+Integration/Feature+Integration+Guides/Gift+Cards+Feature+Integration/nopayment.svg)

> ### Prepare order state machines: Gift Card usage process

Reviewer 2: ```suggestion
### 9) Prepare order state machines: the Gift Card usage process
```

                <event>stock-update</event>
            </transition>

            <transition>
                <source>delivered</source>
                <target>ready for return</target>
                <event>return</event>
            </transition>

            <transition happy="true">
                <source>delivered</source>
                <target>closed</target>
                <event>close</event>
            </transition>

        </transitions>

        <events>
            <event name="authorize" onEnter="true"/>
            <event name="export" onEnter="true" manual="true" command="Oms/SendOrderConfirmation"/>
            <event name="ship" manual="true" command="Oms/SendOrderShipped"/>
            <event name="stock-update" manual="true"/>
            <event name="close" manual="true" timeout="1 hour"/>
            <event name="return" manual="true" />
        </events>
    </process>

    <process name="DummyRefund" file="DummySubprocess/DummyRefund01.xml"/>

</statemachine>
```

![Nopayment](https://spryker.s3.eu-central-1.amazonaws.com/docs/Migration+and+Integration/Feature+Integration+Guides/Gift+Cards+Feature+Integration/nopayment.svg)

> ### Gift Card purchase process

Reviewer 2: Title must start with a verb, in imperative mood—for example:
```suggestion
### 10) Enable the gift card purchase process
```

use Spryker\Zed\Kernel\Container;
use Spryker\Zed\Shipment\ShipmentDependencyProvider as SprykerShipmentDependencyProvider;

class ShipmentDependencyProvider extends SprykerShipmentDependencyProvider
{
    /**
     * @param \Spryker\Zed\Kernel\Container $container
     *
     * @return \Spryker\Zed\Shipment\Dependency\Plugin\ShipmentMethodFilterPluginInterface[]
     */
    protected function getMethodFilterPlugins(Container $container)
    {
        return [
            new OnlyGiftCardShipmentMethodFilterPlugin(),
        ];
    }
}
```

{% info_block warningBox "Verification" %}

Make sure the following:
* You can add a configured gift card product to the cart and purchase it.
* The gift card item doesn't have any discounts applied.
* During the checkout process, shipment method selection is optional if there is only a gift card in the cart.
* The GiftCart/ShipGiftCard OMS command is invoked, and the customer receives an email with the generated gift card code once the order is placed.

Before you can verify these points, you need to complete the [feature frontend integration](#install-feature-frontend).

{% endinfo_block %}

> #### Gift Card code usage process

Reviewer 2: ```suggestion
#### 11) Enable the gift card code usage process
```

2. Apply glossary keys:

```bash
console data:import:glossary
```

{% info_block warningBox "Verification" %}

Make sure that in the database, the configured data has been added to the `spy_glossary` table.

{% endinfo_block %}

### 3) Set up widgets

Register the following global widget:

| WIDGET | SPECIFICATION | NAMESPACE |
| --- | --- | --- |
| CartCodeFormWidget |Provides a cart code activation form.  | SprykerShop\Yves\CartCodeWidget\Widget |

**src/Pyz/Yves/ShopApplication/ShopApplicationDependencyProvider.php**

```php
<?php

namespace Pyz\Yves\ShopApplication;

use SprykerShop\Yves\CartCodeWidget\Widget\CartCodeFormWidget;
use SprykerShop\Yves\ShopApplication\ShopApplicationDependencyProvider as SprykerShopApplicationDependencyProvider;

class ShopApplicationDependencyProvider extends SprykerShopApplicationDependencyProvider
{
    /**
     * @return string[]
     */
    protected function getGlobalWidgets(): array
    {
        return [
            CartCodeFormWidget::class,
        ];
    }
}
```

{% info_block warningBox "Verification" %}

Make sure that the widget is displayed on the **Cart** page and the **Summary** page of the checkout process.

{% endinfo_block %}

> ### 4) Enable controllers

Reviewer 2: If this section have only one sub step, we can merge them or replace the step name with the sub-step name as follows:
```suggestion
### 4) Enable the route list
```

> #### Route List

Reviewer 2: ```suggestion
```
